src/common/Finisher.cc

```diff
@@ -62,8 +62,24 @@ void *Finisher::finisher_thread_entry()
       }
 
       // Now actually process the contexts.
+      int i = 0;
+      uint64_t num_events = 50;
+      uint64_t interval_ns = 5;
+      if (g_ceph_context) {
+        num_events = g_conf().get_val<uint64_t>("finisher_batch_events");
+        interval_ns = g_conf().get_val<uint64_t>("finisher_sleep_interval");
+      }
       for (auto p : in_progress_queue) {
 	p.first->complete(p.second);
+
+	// Give up the CPU to give other threads those are waiting the
```

Comment:
do you mean there could be lots of contexts queued in `finisher_queue` actually can hold, for instance, `mds_lock`, when its `complete()` is called, and hence starve other tasks which also want to acquire the same lock?

Reply:
Yeah, it is.

Reply:
how are "other tasks" scheduled? and i am not sure if it's related to the fairness of scheduling without more context. adding bubbles unconditionally in the queue can hurt the performance.

Reply:
From my test, because the finisher queue has many `complete()` been queued and needed to be run, and each of them will try to acquire/release the `mds_lock` and sometimes they will always succeed prior to other threads, even the other threads was already waiting for the `mds_lock` before some of the `complete()`.

I have tried to 'yield()' the finisher thread, but it will be scheduled back immediately. 

The implementations of the Mutex (e.g. std::mutex in C++) do not guarantee fairness, they do not guarantee that the lock will be acquired by threads in the order that they called lock(). So IMO it will depend on the threads' scheduling policy.

Reply:
> how are "other tasks" scheduled?

could you please share more details on this?

Reply:
also, couple thoughts

- what the thread priorities of these competing threads  ?
- have you tried to use `SCHED_FIFO`?
- if the `complete()` implementations take this grand lock to do long lasting task, i'd suggest avoid this if possible or at least to use a finer grained lock.

Reply:
> > how are "other tasks" scheduled?
> 
> could you please share more details on this?
All the threads, including the finisher thread, in the MDS process are based SCHED_OTHER policy:

`root        9624       1    9624   27 TS   19 19:59 ?        00:00:00 /data/rhcs/ceph/build/bin/ceph-mon -i a -c /data/rhcs/ceph/build/ceph.conf `

Reply:
> also, couple thoughts
> 
> * what are the thread priorities of these competing threads  ?

All the thread priorities are 19.

> * have you tried to use `SCHED_FIFO`?

Do you mean switch to the finisher thread to FIFO ? Or all the threads ?

> * if the `complete()` implementations take this grand lock to do long lasting task, i'd suggest avoid this if possible or at least to use a finer grained lock.

For each `complete()` it only takes 0.001~0.10 seconds, mostly are around 0.01 or less than from my test.

In case when deleting hundreds of directories, in MDS the journal will be flushed in batch and it's possibly hundreds of complete queued, so after the flush finishes, the finisher thread will call all the callbacks, never scheduled out until all the callbacks are ran.

Reply:
> > also, couple thoughts
> > 
> > * what are the thread priorities of these competing threads  ?
> 
> All the thread priorities are 19.
> 
> > * have you tried to use `SCHED_FIFO`?
> 
> Do you mean switch to the finisher thread to FIFO ? Or all the threads ?

the threads which are trying to acquire the `mds_lock`.

> 
> > * if the `complete()` implementations take this grand lock to do long lasting task, i'd suggest avoid this if possible or at least to use a finer grained lock.
> 
> For each `complete()` it only takes 0.001~0.10 seconds, mostly are around 0.01 or less than from my test.
> 
> In case when deleting hundreds of directories, in MDS the journal will be flushed in batch and it's possibly hundreds of complete queued, so after the flush finishes, the finisher thread will call all the callbacks, never scheduled out until all the callbacks are ran.

hmm, this does not sound like a scaleable way to handle large amount of directories removal. lemme think about it.

Reply:
> > > also, couple thoughts
> > > 
> > > * what are the thread priorities of these competing threads  ?
> > 
> > 
> > All the thread priorities are 19.
> > > * have you tried to use `SCHED_FIFO`?
> > 
> > 
> > Do you mean switch to the finisher thread to FIFO ? Or all the threads ?
> 
> the threads which are trying to acquire the `mds_lock`.
> 

Here it mainly for the `DispatchQueue` dispatch thread, I will try that.


> > > * if the `complete()` implementations take this grand lock to do long lasting task, i'd suggest avoid this if possible or at least to use a finer grained lock.
> > 
> > 
> > For each `complete()` it only takes 0.001~0.10 seconds, mostly are around 0.01 or less than from my test.
> > In case when deleting hundreds of directories, in MDS the journal will be flushed in batch and it's possibly hundreds of complete queued, so after the flush finishes, the finisher thread will call all the callbacks, never scheduled out until all the callbacks are ran.
> 
> hmm, this does not sound like a scaleable way to handle large amount of directories removal. lemme think about it.

In cephfs, for many of the client requests, there has two replies, the unsafe and safe ones. For the unsafe one it will just mean that all the operations have been journaled, but not guarantee that the journal logs will be flushed to the disk. And the safe one will be delayed and issued just after the journal logs successfully flushed, in the `complete()` callback.

In the client side, once the unsafe reply is received it will assume it was successful and just returns success to its uplayer.

So, not only for rmdir requests, but also for many other requests will also have the same issue. If there has hundreds of those requests come, so the finisher thread may run all the `complete()` callbacks by holding the `mds_lock` for a long time.

Reply:
> > > also, couple thoughts
> > > 
> > > * what are the thread priorities of these competing threads  ?
> > 
> > 
> > All the thread priorities are 19.
> > > * have you tried to use `SCHED_FIFO`?
> > 
> > 
> > Do you mean switch to the finisher thread to FIFO ? Or all the threads ?
> 
> the threads which are trying to acquire the `mds_lock`.
> 
Here the Finisher thread is mainly blocking the `DispatchQueue` queue thread, which is handling the client side requests, like the `rmdir` request in this issue.

From my test the `DisptchQueue` queue thread will be queued thousands of client requests sometimes, which will also need to acquire the `mds_lock` for each request. This is like what the Finisher thread will do. It will be also a problem, please see https://tracker.ceph.com/issues/51734.

If we set the `SCHED_FIFO` policy for `DispatchQueue` queue thread, shouldn't the thread will run without being scheduled until it finishes all the client requests too ?

This will also cause other thread to starve.

I have test this, it won't work for me.

Reply:
Till now this works well for me and it's very hard to reproduce it with this fixing. 

I am also thinking maybe we can add one flag or record the oldest lock waiter's timestamp before each `mds_lock`, and only if that flag or the timestamp is set will the `Finisher` thread add the bubble then, or the `Finisher` thread will execute all the `complete()`s  without stopping ?

@tchaikov @cbodley @batrick  Will that make sense ?